File: java/leetcode/dp/perfect_squares.java
/**
 * 279. Perfect Squares
 * Given a positive integer n, find the least number of perfect square numbers (for example, 1, 4, 9, 16, ...) which sum to n.

Example 1:

Input: n = 12
Output: 3 
Explanation: 12 = 4 + 4 + 4.
Example 2:

Input: n = 13
Output: 2
Explanation: 13 = 4 + 9.
 */

import java.util.List;
import java.util.ArrayList;

public class perfect_squares {

    public int numSquares(int n) {
        List<Integer> squareList = generateSquareList(n);
        int[] dp = new int[n+1];
        for (int i = 1; i <= n; i++) {
            int min = Integer.MAX_VALUE;
            for (int square : squareList) {
                if (square > i) {
                    break;
                }
                min = Math.min(min, dp[i - square] + 1);
            }
            dp[i] = min;
        }
        return dp[n];
    }

    private List<Integer> generateSquareList(int n) {
        List<Integer> squareList = new ArrayList<>();
        int diff = 3;
        int square = 1;
        while (square <= n) {
            squareList.add(square);
            square += diff;
            diff += 2;
        }
        return squareList;
    }
}

File: golang/leetcode/dp/perfect_squares.go
package dp

import "math"

// 279. Perfect Squares

func numSquares(n int) int {
	squarelist := generateSquares(n)
	dp := make([]int, n+1)
	for i := 1; i <= n; i++ {
		min := math.MaxInt32
		for _, square := range squarelist {
			if square > i {
				break
			}
			min = Min(min, dp[i-square]+1)
		}
		dp[i] = min
	}
	return dp[n]
}

func generateSquares(n int) []int {
	squares := make([]int, 0)
	square, diff := 1, 3
	for square <= n {
		squares = append(squares, square)
		square += diff
		diff += 2
	}
	return squares
}

/*
func Min(x, y int) int {
	if x < y {
		return x
	}
	return y
}
*/
